packages/eslint-plugin/tests/docs.test.ts

```diff
@@ -180,6 +181,32 @@ describe('Validating rule docs', () => {
           }
         });
       }
+
+      test('must include only valid code samples', () => {
+        for (const token of tokens) {
+          if (token.type !== 'code') {
+            continue;
+          }
+
+          const lang = token.lang?.trim();
+          if (!lang?.startsWith('ts')) {
+            return;
+          }
+
+          try {
+            parseForESLint(token.text, {
+              ecmaFeatures: {
+                jsx: lang.startsWith('tsx'),
+              },
+              ecmaVersion: 'latest',
+              sourceType: 'module',
+              range: true,
+            });
+          } catch (e) {
+            throw new Error(`Parsing error:\n\n${token.text}`);
```

Comment:
Nit: a bit more informative-ness in the error if we use `cause`, what do you think?

```suggestion
          } catch (e) {
            throw new Error(`Parsing error:\n\n${token.text}`, { cause: e });
```

Alternately my suggestion would be:

```suggestion
          } catch {
            throw new Error(`Parsing error:\n\n${token.text}`);
```

...but my guess is showing the caught error would be nicer.

Reply:
What about this?

```ts
try {
  // ...
} catch (e) {
  if (
    e &&
    typeof e === 'object' &&
    'lineNumber' in e &&
    typeof e.lineNumber === 'number' &&
    'column' in e &&
    typeof e.column === 'number' &&
    'message' in e
  ) {
    const codeFrameLines = token.text.split('\n');
    codeFrameLines.splice(
      e.lineNumber,
      0,
      ' '.repeat(e.column) + '^: ' + e.message,
    );
    throw new Error(`Parsing error:\n\n${codeFrameLines.join('\n')}`);
  }
  throw new Error(`Parsing error:\n\n${token.text}`);
}
```

The code looks pretty scary, but what it does - it checks if `e` is `TSError`

https://github.com/typescript-eslint/typescript-eslint/blob/c61c3eedf5dec98c1548c8ab533dcbb756a9cd7b/packages/typescript-estree/src/node-utils.ts#L688

And if `e` looks like `TSError` it prints code frame with pointer on the location of parsing error: 

```
  ● Validating rule docs › consistent-type-exports.md › must include only valid code samples

    Parsing error:

    interface ButtonProps {
      onClick: () => vo id;
                        ^ ';' expected.
    }

    class Button implements ButtonProps {
      onClick = () => console.log('button!');
    }

    export { Button, ButtonProps };

      219 |                 ' '.repeat(e.column) + '^ ' + e.message,
      220 |               );
    > 221 |               throw new Error(`Parsing error:\n\n${codeFrameLines.join('\n')}`);
          |                     ^
      222 |             }
      223 |             throw new Error(`Parsing error:\n\n${token.text}`);
      224 |           }

      at Object.<anonymous> (tests/docs.test.ts:221:21)
```

(We can't import `TSError` from `@typescript-eslint/typescript-estree` because of `no-typescript-estree-import` internal rule)

Reply:
That seems like a lot of effort for something that doesn't happen very often & has a pretty clear error by default 😄

Reply:
Yup, you're right, but we have a little issue with `cause`:

It's defined in `ES2022`: https://github.com/microsoft/TypeScript/blob/01a51d2d017dd50d487835282b7e4a2a4f7300ab/src/lib/es2022.error.d.ts,
but we have `ES2021`:
https://github.com/typescript-eslint/typescript-eslint/blob/78db0add16e9d7ea23c5ffc13341cc13b16eec2d/tsconfig.base.json#L17

Soo..

```
$ yarn tsc --noEmit
tests/docs.test.ts:206:64 - error TS2554: Expected 0-1 arguments, but got 2.

206             throw new Error(`Parsing error:\n\n${token.text}`, { cause: e });
                                                                   ~~~~~~~~~~~~


Found 1 error in tests/docs.test.ts:206
```

I guess we stick with your second suggestion :slightly_smiling_face: